Bluetooth: host: Extended advertising fixes

Tronil opened the pull request:
Fix missing endianness conversion for props in le_ext_adv_param_set

Fix advertising failing to resume with BT_ADV_PERSIST set in
bt_hci_le_adv_set_terminated due to BT_ADV_ENABLED not getting
cleared before bt_le_adv_resume gets called

Signed-off-by: Troels Nilsson <trnn@demant.com>

Tronil committed:
Bluetooth: host: Fix missing endianness conversions

Fix missing endianness conversion for props in le_ext_adv_param_set
and bt_le_per_adv_set_param

Signed-off-by: Troels Nilsson <trnn@demant.com>

Thalley commented on subsys/bluetooth/host/adv.c:
```diff
@@ -1192,6 +1192,7 @@ static int le_ext_adv_param_set(struct bt_le_ext_adv *adv,
 
 	cp->sid = param->sid;
 
+	cp->props = sys_cpu_to_le16(cp->props);
```
Two comments:
1) The same issue appears to be in `bt_le_per_adv_set_param`
2) I think it would be more logical if we, instead of storing the properties directly as CPU endianess in `cp->props` (like `cp->props |= BT_HCI_LE_ADV_PROP_TX_POWER;`), we either
A) Add a local variable, `uint16_t props`, and then store them in that and then do `cp->props = sys_cpu_to_le16(props)` to be most consistent with the other `cp` assignments, or find a way store the bits directly as little endian (can you do `cp->props |= sys_cpu_to_le16(BT_HCI_LE_ADV_PROP_TX_POWER);`?)

Tronil replied:
Sure - I'll look into that

Tronil replied:
Changed to use a local `props` variable and added fix to `bt_le_per_adv_set_param`

Thalley commented on subsys/bluetooth/host/adv.c:
```diff
@@ -1896,18 +1897,37 @@ void bt_hci_le_adv_set_terminated(struct net_buf *buf)
+	}
+
+#if (CONFIG_BT_ID_MAX > 1) && (CONFIG_BT_EXT_ADV_MAX_ADV_SET > 1)
+	was_adv_enabled = atomic_test_bit(adv->flags, BT_ADV_ENABLED);
+#endif
+
+	atomic_clear_bit(adv->flags, BT_ADV_ENABLED);
+
 #if (CONFIG_BT_ID_MAX > 1) && (CONFIG_BT_EXT_ADV_MAX_ADV_SET > 1)
 	bt_dev.adv_conn_id = adv->id;
 	for (int i = 0; i < ARRAY_SIZE(bt_dev.cached_conn_complete); i++) {
 		if (bt_dev.cached_conn_complete[i].valid &&
 		    bt_dev.cached_conn_complete[i].evt.handle == evt->conn_handle) {
-			if (atomic_test_bit(adv->flags, BT_ADV_ENABLED)) {
+			if (was_adv_enabled) {
```
So this change is to make connectable extended advertising persistable? 

Is this fully working? 

In `bluetooth.h` we have 
```c
	/**
	 * @brief Advertise one time.
	 *
	 * Don't try to resume connectable advertising after a connection.
	 * This option is only meaningful when used together with
	 * BT_LE_ADV_OPT_CONNECTABLE. If set the advertising will be stopped
	 * when bt_le_adv_stop() is called or when an incoming (peripheral)
	 * connection happens. If this option is not set the stack will
	 * take care of keeping advertising enabled even as connections
	 * occur.
	 * If Advertising directed or the advertiser was started with
	 * @ref bt_le_ext_adv_start then this behavior is the default behavior
	 * and this flag has no effect.
	 */
	BT_LE_ADV_OPT_ONE_TIME = BIT(1),
```

specifically because extended advertising is not persistable. 

Besides that, we can have multiple connectable advertising sets, and just having one set taking exclusive rights on being connectable like here, perhaps isn't what we want. See the issue I raised in https://github.com/zephyrproject-rtos/zephyr/issues/30699.

Tronil replied:
Actually, this was to fix legacy advertising when using the extended advertising HCI interface. A lot of our tests started failing once we enabled CONFIG_BT_EXT_ADV/CONFIG_BT_CTLR_ADV_EXT while still only using legacy advertising.

But if it is specifically not supported when using extended advertising in the HCI, then I guess we will have to live with it.

Tronil replied:
> HCI does not have the concept of persistable advertising at all - That is a Zephyr feature.
> 
> If you are not using `bt_le_ext_adv_start` directly, persisting advertising should still work, but I'll have to try out the BT shell or something to determine if there's indeed a problem and then do a proper review of this change.

Sorry I was probably a bit unclear. We are using `bt_le_adv_start()` (without BT_LE_ADV_OPT_ONE_TIME) to start advertising and resume advertising after a connection. After turning on the extended advertising support it no longer resumes the advertising though.

My fix does not take this bit into account though:
`	 * If Advertising directed or the advertiser was started with
	 * @ref bt_le_ext_adv_start then this behavior is the default behavior
	 * and this flag has no effect.
`

I'm not sure how we can determine when advertising has been started by a call to `bt_le_ext_adv_start()` vs `bt_le_adv_start()` though.

Thalley replied:
> I'm not sure how we can determine when advertising has been started by a call to bt_le_ext_adv_start() vs bt_le_adv_start() though.

`bt_le_adv_start()` will always use the `bt_dev.adv` advertiser, and I think that the `resume` function will just use that. 

> After turning on the extended advertising support it no longer resumes the advertising though.

OK, so this change is to fix resuming when extended advertising is enabled, but when still using `bt_le_adv_start()` - That makes sense. I'll have to run a few tests locally to see if there's indeed a general issue (none has reported before AFAIK), and it has been working before IIRC.

Tronil replied:
> > I'm not sure how we can determine when advertising has been started by a call to bt_le_ext_adv_start() vs bt_le_adv_start() though.
> 
> `bt_le_adv_start()` will always use the `bt_dev.adv` advertiser, and I think that the `resume` function will just use that.

Ah yes as far as I can tell `bt_le_adv_resume()` will only resume the legacy advertising (and only when relevant). So it should be safe to call it the way this fix does.

Thalley replied:
@Tronil Can you elaborate on why/when the advertising isn't resumed? I just used the shell on `main` with `CONFIG_BT_EXT_ADV=y` and where `bt_le_adv_start` calls `bt_le_adv_start_ext`, and where `bt_le_adv_resume` is called on disconnect and thus calling `bt_le_adv_set_enable` again. 

So to me it seems to work as intended.

Tronil replied:
> @Tronil Can you elaborate on why/when the advertising isn't resumed? I just used the shell on `main` with `CONFIG_BT_EXT_ADV=y` and where `bt_le_adv_start` calls `bt_le_adv_start_ext`, and where `bt_le_adv_resume` is called on disconnect and thus calling `bt_le_adv_set_enable` again.

The problem isn't on disconnect - rather, advertising should be resumed after a connection has been established, so we can establish several connections at the same time. I would expect it to resume after a disconnect since BT_ADV_ENABLED has then long since been cleared - however, that is too late.

Thalley commented on subsys/bluetooth/host/adv.c:
```diff
@@ -1896,18 +1900,37 @@ void bt_hci_le_adv_set_terminated(struct net_buf *buf)
+	       evt->status, evt->adv_handle, conn_handle,
+	       evt->num_completed_ext_adv_evts);
+
+	if (!adv) {
+		BT_ERR("No valid adv");
+		return;
+	}
```
This check should be done before `(void)bt_le_lim_adv_cancel_timeout(adv);`, shouldn't it?

Thalley reviewed (changes requested):
Minor comment that needs to be addressed, otherwise LGTM.

jhedberg commented:
Looks fine to me, but the fact that you're using the plural "fixes" makes me think perhaps this needs to be split into multiple commits?

cvinayak commented:
> > Looks fine to me, but the fact that you're using the plural "fixes" makes me think perhaps this needs to be split into multiple commits?
> 
> You could split it into one PR for the missing endianness conversion and one for the advertisement not resuming properly I guess. They are just so small that I think it will be easier on everyone to fix it in one PR?

Endianness changes does have quite a few lines changed and is a separate bug fix in itself. Please have two commits , one for endianness bug fix and the other for legacy advertising persistence with suitable commit titles, in this single PR. (Each small commit that is buildable and testable aids use of git bisect when need be).

cvinayak reviewed (dismissed):
Changes look good, requesting split into two different commits.

jhedberg commented:
> You could split it into one PR for the missing endianness conversion and one for the advertisement not resuming properly I guess. They are just so small that I think it will be easier on everyone to fix it in one PR?

Yes, one PR is fine. Note that I didn't request multiple PRs, just multiple commits.

Tronil committed:
Bluetooth: host: Fix resume failing with extended advertising

Fix advertising failing to resume with BT_ADV_PERSIST set in
bt_hci_le_adv_set_terminated due to BT_ADV_ENABLED not getting
cleared before bt_le_adv_resume gets called

Signed-off-by: Troels Nilsson <trnn@demant.com>

Tronil commented:
Split into two commits, one for the missing endianness conversions and one for the failing resume of advertising

Tronil commented:
> > Split into two commits, one for the missing endianness conversions and one for the failing resume of advertising
> 
> Compliance check failed due to mismatch in signed-off-by. Use `git commit --amend -s ` to add Signed-off-by to your top commit.

Doh - should be fixed now
